allennlp/modules/token_embedders/pretrained_transformer_embedder.py

```diff
@@ -119,6 +126,22 @@ def __init__(
             self._scalar_mix = ScalarMix(self.config.num_hidden_layers)
             self.config.output_hidden_states = True
 
+        # Optionally, re-initialize the parameters of certain layers.
+        self._reinit_layers = cast(List[int], reinit_layers)
+        if self._reinit_layers and load_weights:
+            num_layers = len(self.transformer_model.encoder.layer)
```

Comment:
This should work fine for BERT, but won't for other models like XLM that name their modules differently. In general I think it'll be brittle to rely on guessing the internal module structure of transformer models.

But I can think of alternative which - although it's a little less friendly from a user perspective - still provides this useful functionality:
- Change the type of `reinit_layers` to `Optional[Union[int, List[int], List[str]]]`.
- When `reinit_layers` is `List[str]`, we interpret it as a list of regular expressions and we reinitialize all modules that match any of the regular expressions in the list.
- Otherwise when `reinit_layers` is `int` or `List[int]` we try the current approach, but if the given `transformer_model` does not have this module structure we throw a `ConfigurationError` and suggest that the user use the `List[str]` form of this parameter instead.

Reply:
Okay, I took a crack at this.

- I renamed `reinit_layers` to `reinit_modules` just because re-initializing modules (instead of whole layers) is possible when it is a list of regexes.
- Updated the type of the argument to `Optional[Union[int, List[int], List[str]]]`
- A configuration error is thrown if the model's layers cannot be easily indexed. In practice, most models I tried from hugging face could (BERTs, RoBERTa's) but you were right that XLM based models cant.
- Added tests for when `reinit_modules` is a list of strings.

Unfortunately, I cant seem to get the weights to re-initialize. I figured that after finding a match between the regex and some `module` name I could just call `module.apply(self.transformer_model._init_weights)` like so

```python
for regex in self._reinit_modules:
    for name, module in self.transformer_model.named_modules():
        if re.search(regex, name):
            module.apply(self.transformer_model._init_weights)
```

but this doesn't actually re-init the `module`s weights? Strange. I will have to dig into the HF repo and docs to figure out why.